### PR TITLE
973: Implement Table of Contents block

### PR DESCRIPTION
**Pull request by kadamwhite**
Uses Group block filter results to build array of ToC items, then saves that in-post as a nav UL. For [WIKI-973](https://github.com/humanmade/wikimedia/issues/973)

This block parses the other blocks on the page to identify waypoint-labeled groups (see #11 / [WIKI-963](https://github.com/humanmade/wikimedia/issues/963)), then saves those anchors as an on-page UL element. It does not currently provide any styling.

**Editor**

toc-demo-editor

Clicking a link in the ToC within the editor will jump to that group.

**Frontend**

image

### CODE REVIEW

**Review by philipjohn — APPROVED** [review]
This works really nicely, great stuff!

File: src/blocks/table-of-contents/edit.js [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	}, [ waypoints, setAttributes ] );
+
+	return (
+		<ul { ...useBlockProps() }>
+			{ waypoints.map( ( waypoint ) => (
+				<li key={ `waypoint-${ waypoint.clientId }` }>
+					{ /* eslint-disable-next-line jsx-a11y/anchor-is-valid -- Maintains styling parity, we save this as an <a>. */ }
+					<a
+						href={ `#${ waypoint.tocSlug }` }
+						onClick={ () => {
```
Comment (goldenapples) [inline src/blocks/table-of-contents/edit.js:89]: This is really nice!

**Review by goldenapples — APPROVED** [review]
This works really well.

I imagine we'll need a view script for this block as well, to implement smooth scrolling and set an offset to the scroll targets to handle the sticky header. But that can be done while implementing the block front end.

**Comment by kadamwhite** [comment]
Thanks for the review, @goldenapples ! Acknowledged on the styling needs. I'm ambivalent about smooth scrolling, that's up to Ivan / Sara / Olga, but we can follow it up later on when we add the whole desktop and mobile styling. I'll merge this for now to anchor the core functionality.

**Pull request merged**